### ISSUE TITLE
CLI Lcov coverage when running html runner

### ISSUE THREAD

**Issue by ICTechnology** [mmanela/chutzpah#640]
Hello,

I'm not 100% sure this is possible.

I'll explain my problem. We currently use QUnit to test our JS, we use individual test runners to do this, with a big "MasterTestRunner" which includes all of our test files and their dependencies.

Currently i'm trying:

`chutzpah.console.exe AlertsTestRunner.html /coverage /lcov cov`

This runs the tests and I get 5 total ran, but the cov file is empty. Now if I point it at the JsTest file, the tests run, but fail (due to it not know what modules are etc) BUT it does give me a populated lcov file. Not 100% sure if it's legit.

Currently our setup is we have a Powershell script that kicks off  node to run node-qunit-phantomjs against our MasterTestRunner. This works brilliantly, runs all the tests and gives us exit codes, but our last piece is to have a lcov file output, which we can then feed in to SonarQube to report on our JS Test coverage.

Basically I'm asking for a bit of help, is this possible? If so could you point me in the right direction please?

Many Thanks,

Iain

**Comment by mmanela** [comment]
Chutzpah needs to run on a JS file / chutzpah.json file to be able to generate coverage since it needs to be able to create its own HTML file. If you set up Chutzpah (through its chutzpah.json file) to correctly run directly on your JS files then you should be able to get coverage.

**Comment by ICTechnology** [comment]
Hello,

Our problem is we use a HTML test runner to set up and inject markup, which the JS then manipulates. I take it then there isn't a way we can do this with our current setup ? The JS files just die because their dependencies are included in the test runner.

**Comment by mmanela** [comment]
Chutzpah lets you reference HTML files and they will be injected for you into the test harness

**Comment by ICTechnology** [comment]
You mean to change it so the JS test file, references a HTML file for it's markup and dependencies ?

**Comment by mmanela** [comment]
Yes, let the JS files (or chutzpah.json) drive it.

**Comment by jwaliszko** [comment]
I've just came across similar issue and referenced the html file from the JS file:
`/// <reference path="./form.tests.html" />`.

Without the instrumentation it works perfect.
When coverage is on, the test results are sometimes affected. I've described the issue and attached sample which allows to reproduce the error here: https://github.com/alex-seville/blanket/issues/572.

@mmanela could you please take a look?
Maybe I've made some configuration mistake which affects the results.

Thanks for Chutzpah. It's indispensable tool for me.

**Comment by mmanela** [comment]
Try this as your chutzpah.json

```
{
  "framework": "qunit",
  "CodeCoverageExcludes": ["*jquery*"],
  "References": [
    { "Path": "extern/jquery-1.8.2.js", "IsTestFrameworkFile": true},
    { "Path": "extern/jquery.validate-1.10.0.js", "IsTestFrameworkFile": true},
    { "Path": "extern/jquery.validate.unobtrusive-3.1.1.js", "IsTestFrameworkFile": true},
    { "Path": "expressive.annotations.validate.js", "IsTestFrameworkFile": true },
    { "Path": "form.tests.html" },
  ]
},

"Tests": [
  { "Path": "form.tests.js" }
]
```

**Comment by jwaliszko** [comment]
Execution of `chutzpah.console.exe /path form.tests.js /coverage` command, driven by the *chutzpah.json* file shown above, results in the coverage of *form.tests.js* solely, which is not what I'd like to get:

![image](https://user-images.githubusercontent.com/1596462/28252419-a05d12dc-6a92-11e7-8df0-5b03892b4592.png)

The aim is to collect the coverage of *expressive.annotations.validate.js* instead.

If `IsTestFrameworkFile` is switched to `false`, i.e.
`{ "Path": "expressive.annotations.validate.js", "IsTestFrameworkFile": false }`

the same issue appears as presented in my primary solution, i.e. unit test failure:

![image](https://user-images.githubusercontent.com/1596462/28252415-8fee1c98-6a92-11e7-86c3-2fdd5dbdb5dd.png)

**Comment by mmanela** [comment]
I see, it may depends on the life-cycle of jQuery validate Blanket (when instrumenting a file) takes it out of its normal order and load a new instrumented version. If a library is assuming things exist at a certain time it can mess this up. In this case I do not know the root issue but I have seen this class before.

In the future I plan to switch chutzpah to another library like Istanbul, which works as a pre-step and should alleviate this issue but that won't be soon.

**Comment by jwaliszko** [comment]
@mmanela: is switching to pre-step alternative on your schedule, or still not really on the horizon for now?

**Comment by mmanela** [comment]
I would love to but not something I have time to work on.
